doc/Subprocess debugging.md

```diff
@@ -0,0 +1,318 @@
+
+Debuggee_1 -->>- Adapter: confirm "terminate"
+
+Debuggee_1 -X- Debuggee_1: exits
+
+Adapter -->>- IDE: confirm "disconnect" from Debuggee_1
+
+Adapter -X Adapter: exits
+```
+
+
+
+## "attach" scenario
+
+1. User starts debuggee process with debug server in it (ptvsd command line or `ptvsd.enable_attach()`).
+1. User starts debugging (F5) with "attach" debug config.
+1. User code spawns child process.
+1. User disconnects from debuggee.
+1. User reconnects to debuggee.
+
+```mermaid
+sequenceDiagram
+# Install "GitHub + Mermaid" from the Chrome Web Store to render the diagram
+
+participant IDE
+participant Adapter
+participant Debuggee_1
+participant Debuggee_2
+
+
+Note left of Debuggee_1: user spawns<br/>debuggee
+
+Debuggee_1 ->>+ Adapter: spawn and pass adapter listener port (cmdline)
+
+Adapter -->>- Debuggee_1: connect to adapter listener port
+
+Adapter ->> Debuggee_1: request "setDebuggerProperty" for server listener port
+
+
+Note left of IDE: user starts<br/>debugging
+
+IDE ->> Adapter: connect to IDE listener port
+
+IDE ->>+ Adapter: request "attach"
+
+Adapter ->>+ Debuggee_1: request "initialize", "attach"
+
+Debuggee_1 -->>- Adapter: respond to "initialize", "attach"
+
+Debuggee_1 ->>+ Debuggee_1: start debug session
+
+Adapter -->>- IDE: respond to "attach"
+
+loop every message between IDE and Debuggee_1
+Note over IDE,Debuggee_1: propagate message
+end
+
+
+Note right of Debuggee_1: user code spawns<br/>child process
+
+Debuggee_1 ->>+ Debuggee_2: spawn and pass server listener port (cmdline)
+
+Debuggee_2 ->>- Adapter: connect to server listener port
+
+Adapter ->>+ Debuggee_2: request "pydevd_systemInfo"
+
+Debuggee_2 -->>- Adapter: respond to "pydevd_systemInfo"
+
+Adapter ->>+ IDE: "ptvsd_subprocess" event
+
+IDE ->>- Adapter: connect to IDE listener port
+
+IDE ->>+ Adapter: request "attach" to Debuggee_2
+
+Adapter ->>+ Debuggee_2: request "initialize", "attach"
+
+Debuggee_2 -->>- Adapter: respond to "initialize", "attach"
+
+Debuggee_2 ->>+ Debuggee_2: start debug session
+
+Adapter -->>- IDE: respond to "attach"
+
+loop every message between IDE and Debuggee_2
+Note over IDE,Debuggee_2: propagate message
+end
+
+
+Note left of IDE: user detaches IDE
+
+IDE ->>+ Adapter: request "disconnect" from Debuggee_2
+
+Adapter ->>+ Debuggee_2: request "disconnect"
+
+Debuggee_2 ->>- Adapter: confirm "disconnect"
+
+Debuggee_2 ->>- Debuggee_2: end debug session
+
+Note over Adapter,Debuggee_2: TCP connection is maintained
+
+Adapter -->>- IDE: confirm "disconnect" from Debuggee_2
+
+IDE ->>+ Adapter: request "disconnect" from Debuggee_1
+
+Adapter ->>+ Debuggee_1: request "disconnect"
+
+Debuggee_1 ->>- Adapter: confirm "disconnect"
+
+Debuggee_1 ->>- Debuggee_1: end debug session
+
+Note over Adapter,Debuggee_1: TCP connection is maintained
+
+Adapter -->>- IDE: confirm "disconnect" from Debuggee_1
+
+Note over Adapter: continues running
+
+
+
+Note left of IDE: User re-attaches IDE<br/>(same host/port)
```

Comment:
Would be nice to say what happens if a new subprocess is created in the meanwhile (say a `Debuggee_3`) 

i.e.: we could do an attach to pid for new processes found (or we could just initially ignore those).

Reply:
The adapter still remains connected to all the processes, so any new ones will also connect to it as they spawn. Since there's no IDE, the adapter won't report "ptvsd_subprocess" at that point, but it will maintain the connection and the session. When the IDE reconnects later, "ptvsd_subprocess" for every subprocess that the adapter is currently aware of will be sent to the IDE, and it will then open additional connections to the adapter for each.

Similarly, if an existing subprocess exits after detach but before re-attach, the IDE doesn't receive "ptvsd_subprocess" for it when it re-attaches. 

This probably needs a separate diagram to show both clearly.